### PR TITLE
feat: [CDS-86005] Add logic for saving K8s Traffic Routing Metadata in Release

### PR DESCRIPTION
**Pull request by alebuh**
## Describe your changes
We need to persist information about Traffic Shifting resources that we create to later update it in the Swap step or TrafficRouting Step (when config type is inherited). As in both cases we are going to update just destinations, we don't need to update all resources, just those where we configure destination rules. For the SMI provider, this is the TrafficSplit resource and for the Istio provider, it's the VirtualService resource.
Update of these resources we are going to perform using a patch function for which we need the following information:
1. Api version
2. Kind (in plural ex: trafficsplits, virtualservices)
3. Resource name

The idea of logic is to go through all traffic routing resources that we have created, find those that contain destination configuration (TrafficSplit and VirtualService), and extract the required information from them to create TrafficRoutingInfoDTO. This object we are going to save in release so it can be easily retrieved when times come
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/57046)

### CODE REVIEW

**Comment by alebuh** [comment]
trigger uts

File: 960-api-services/src/main/java/io/harness/k8s/releasehistory/K8sReleaseConstants.java [hunk 1]
```diff
@@ -35,6 +35,9 @@ public class K8sReleaseConstants {
   public static final String RELEASE_SECRET_RELEASE_COLOR_KEY = "color";
   public static final String RELEASE_SECRET_RELEASE_BG_ENVIRONMENT_KEY = "harness.io/bg-environment";
   public static final String RELEASE_SECRET_RELEASE_MANIFEST_HASH_KEY = "harness.io/manifest-hash";
+  public static final String RELEASE_SECRET_TRAFFIC_ROUTING_INFO_NAME = "harness.io/traffic-routing-info/name";
```
Comment (abosii-harness) [inline 960-api-services/src/main/java/io/harness/k8s/releasehistory/K8sReleaseConstants.java:38]: this wouldn't work, you can a dns domain as prefix and / but not multiple /
Can we just store everything as one annotation and just following a specific format? Or we can just use json

**Review by abosii-harness — CHANGES_REQUESTED** [review]

File: 960-api-services/src/main/java/io/harness/k8s/releasehistory/K8sReleaseConstants.java (hunk 1 above)
Comment (alebuh) [reply 960-api-services/src/main/java/io/harness/k8s/releasehistory/K8sReleaseConstants.java]: done

**Comment by alebuh** [comment]
trigger uts

File: 960-api-services/src/main/java/io/harness/k8s/releasehistory/K8sLegacyRelease.java [hunk 2]
```diff
@@ -43,6 +43,9 @@ public class K8sLegacyRelease implements IK8sRelease {
   private String helmChartVersion;
   private String helmChartRepoUrl;
   private String helmChartSubChartPath;
+  private String trafficRoutingInfoName;
```
Comment (abosii-harness) [inline 960-api-services/src/main/java/io/harness/k8s/releasehistory/K8sLegacyRelease.java:46]: Wouldn't it better to put it a single object?

Comment (alebuh) [reply 960-api-services/src/main/java/io/harness/k8s/releasehistory/K8sLegacyRelease.java]: I was not sure if this object should be serialized/deserialized anywhere so I followed the same format as it's for helm chart info, but I think you are right, we can keep it as a single object.
Changed now

**Review by abosii-harness — APPROVED** [review]

**Comment by alebuh** [comment]
trigger uts

File: 960-api-services/src/main/java/io/harness/k8s/releasehistory/K8sRelease.java [hunk 3]
```diff
@@ -187,6 +190,33 @@ public HelmChartInfoDTO getHelmChartInfo() {
     return K8sReleaseSecretHelper.getHelmChartInfo(releaseSecret);
   }
 
+  @Override
+  public void setTrafficRoutingInfo(TrafficRoutingInfoDTO trafficRoutingInfo) {
+    if (trafficRoutingInfo != null) {
+      try {
+        String trafficRoutingInfoJson = new ObjectMapper().writeValueAsString(trafficRoutingInfo);
+        K8sReleaseSecretHelper.putAnnotationsItem(
```
Comment (abhinavsingh1196) [inline 960-api-services/src/main/java/io/harness/k8s/releasehistory/K8sRelease.java:198]: will it be beneficial to compress this JSON (and b64 encode) and then store it? k8s secrets with declarative rollback store the release manifests also, so we need as much space for them as possible. 
Although, if this object `TrafficRoutingInfoDTO` is not very huge, so there may not be much benefit. But if it grows in future, then compressing it later will need to handle backward compatibility. In that case it's best to do it now itself

Comment (alebuh) [reply 960-api-services/src/main/java/io/harness/k8s/releasehistory/K8sRelease.java:198]: I don't think it's worth doing it, as I don't expect this object to grow. We were thinking of storing resource yaml as encoded into release, but this came as a better solution.

**Comment by alebuh** [comment]
trigger ut5,ut6,ut7,ut9

**Comment by alebuh** [comment]
trigger ut5,ut7,ut9

**Comment by alebuh** [comment]
trigger uts

File: 960-api-services/src/main/java/io/harness/k8s/releasehistory/K8sRelease.java (hunk 3 above)
Comment (abhinavsingh1196) [reply 960-api-services/src/main/java/io/harness/k8s/releasehistory/K8sRelease.java:198]: In that case, we can skip doing this

**Review by abhinavsingh1196 — APPROVED** [review]

**Review by markoharness — APPROVED** [review]

**Review by TarunUba1 — APPROVED** [review]

**Review by GoelPratyush99 — APPROVED** [review]

**Comment by alebuh** [comment]
trigger ut0,ut5,ut7,ut9

**Comment by alebuh** [comment]
trigger ut7,review

**Pull request merged**